hack/update-cluster

```diff
@@ -0,0 +1,81 @@
+#!/bin/sh
+set -e
+
+# Configuration
+#
+# Path to kubeconfig
+KUBECONFIG=${KUBECONFIG:-cluster/auth/kubeconfig}
+# Version of Kubernetes to upgrade to.
+NEW_VERSION=${NEW_VERSION:-"v1.3.0-beta.2_coreos.0"}
+
+# Shortcut for kubectl command
+kc="kubectl --kubeconfig=${KUBECONFIG} --namespace=kube-system"
+# Temp dir for manifests
+temp=$(mktemp -d -t k8s-upgrade-$NEW_VERSION.XXX)
+# New image to deploy.
+new_image=${new_image:-"quay.io\/coreos\/hyperkube:${NEW_VERSION}"}
+
+update_image() {
+  local filepath=$1
+  local newfile=$(echo $filepath | sed "s/\.yaml/-${NEW_VERSION}.yaml/")
+  echo "newfile: " $newfile
+  sed "s/image:.*hyperkube.*$/image: ${new_image}/g" "${filepath}" > "${newfile}"
+}
+
+# Update kube-apiserver
+# Object is a DaemonSet, so we must first delete, then create to update.
+echo
+echo "Updating apiserver"
+$kc get ds kube-apiserver -oyaml > "${temp}/kube-apiserver.yaml"
+update_image "${temp}/kube-apiserver.yaml"
+# Delete kube-apiserver and wait for temp-apiserver to take over.
+$kc delete --timeout=1s -f "${temp}/kube-apiserver.yaml"
```

Comment:
What if you deploy an api-server that "runs" but is non-functional? This will blow away all api-servers at the same time, deploy the non-functional api-server which will then be checkpointed and you're in an unrecoverable state.

Ideally we provide a mechanism for rolling updates for daemonsets (until this is implemented upstream). One option would be to deploy daemonsets side-by-side and using node-labels to coordinate which version is running where.

E.g. node-label: kubelet=1.1.1 vs node-label: kubelet=1.1.2 -- the upgrade tool slowly changes each node to the new version and lets the daemonset controllers coordinate killing / creating the correct pods.

Or another option would be to delete with `--cascade=false`, create the new daemonset, then have this tool kill each pod one-by-one. Before killing each pod it could check that the last update was successful -- if at any point it is not it can be stopped and replaced with the original daemonset.

Less risky because we're only updating 1 pod at a time -- but still risky in that if a pod is deleted for another reason it will be re-created with new manifest (not us controlling the update).

See some upstream discussion about this:

https://github.com/kubernetes/kubernetes/issues/15310
https://github.com/kubernetes/kubernetes/issues/22543